designs/2019-public-config-array-api/README.md

```diff
@@ -0,0 +1,85 @@
+- Start Date: 2019-12-09
+- RFC PR: <!-- leave this empty, to be filled in later -->
+- Authors: @manuth
+
+# Public `ConfigArray` API
+
+## Summary
+
+Allow 3rd-party developers to determine which configuration-files are considered by `eslint`.
+
+## Motivation
+
+In some cases people might want to automatically trigger a new linting process as soon as the configuration of eslint has changed.
+This is a key feature for language servers which provide warnings and fixes to IDEs.
+
+Therefore the language-server needs to know which config-files are considered by `eslint`.
```

Comment:
Humh...
I totally get your point - adding such a method to the public API, in this case, wouldn't add the desired ability of knowing which files to watch.

`import {} from "source";`, `import "source";` and `import var = require("source");`, btw, affect `require.cache` as well so clearing or checking the cache would work out pretty well also for ES modules.

Anyway to reach the goal I'm after you'd have to import the plugins and settings and check `require.cache` for changes, right?

Sounds like this would require another separate method other than `getConfigArrayForFile`?

Reply:
> btw, affect require.cache as well so clearing or checking the cache would work out pretty well also for ES modules.

It's only if the `"source"` was CJS. If the `"source"` is ESM, Node.js caches it only in the internal cache and we cannot access it. The `require.cache` doesn't include imported ESM files.

> Anyway to reach the goal I'm after you'd have to import the plugins and settings and check require.cache for changes, right?

Currently, ESLint doesn't support ESM, so it may work. However, after we supported ESM, it's going to not work properly.

> Sounds like this would require another separate method other than getConfigArrayForFile?

Yes. I think `getConfigArrayForFile` is not useful for this purpose.

Reply:
Alright - so cleaning the cache is of no use as it will stop working at some point anyway...?

Forking a new process might be of no use, too, as this would draw me unable to get the result from `eslint`... Or is there some way to pass JavaScript-objects from/to forked processes or any other alternative?

> Yes. I think `getConfigArrayForFile` is not useful for this purpose.
Fine - shall we cover both the `getConfigArrayForFile` and the language-server stuff in this RFC?